### ISSUE TITLE
Generic Foreign Key add button

### ISSUE THREAD

**Issue by solace** [yourlabs/django-autocomplete-light#1205]
Hello,

I'm not sure if I missed this in another ticket or elsewhere in the documentation, but I am attempting to use a GFK dropdown in admin and would like to be able to have an add button. I understand that a GFK makes this non-explicit which is probably why the example in select2_generic_foreign_key does not have an add button, but is there a way to specify an add endpoint for a particular model? Or do I need to use the AddAnother library to do this for GFK?

Thanks

**Comment by jpic** [comment]
Great question!

I reverse engineered the system and have a write up here that should be
useful :
https://yourlabs.org/posts/2012-03-27-howto-javascript-popup-form-returning-value-for-select-like-django-admin-for-foreign-keys/

I suppose django-add-another should support it, but basically you'll need
the pop-up to return the composite key of ctypeid-objid and then it the
rest should work as usual (you need the current master of
django-querysetsequence though, unless the last release now embeds the
recent contributions made there).

If you have trouble then please reproduce in the test project so that we
can take a look and eventually finish up by squashing the remaining bug if
any.

Le mer. 30 déc. 2020 à 07:35, Michele Ong <notifications@github.com> a
écrit :

**Comment by solace** [comment]
Thanks for the tips. I ended up with the solution in this gist using `django-addanother`:
https://gist.github.com/solace/d3ef0050c22bfe26b5a2eae842006704

Summary:

1. Extend `AddAnotherWidgetWrapper` to set `self.choices` otherwise the wrapped widget will raise an error about missing this attribute.
2. Extend the ModelAdmin for the model that needs to be addable to allow popup response to return `ctypeid-objid`.
3. Extend the model form `__init__` to wrap the widget with `AddAnotherWidgetWrapper`. Wrapping in the form Meta class does not work as described in the `django-addanother` documentation.

I did install `django-querysetsequence` at master while I was experimenting, but I'm not actually sure if that was necessary with this solution.

Not sure if there was a way to do this without `django-addanother` but if this gives you ideas, I hope it helps. :)
